### PR TITLE
nixos/hidpi: Harmonise default with documented recommendations

### PR DESCRIPTION
**Pull request by nicoonoclaste**
###### Description of changes

In the NixOS `hidpi` module:
- Disable anti-aliasing and font hinting, as per documented recommendations for hiDPI screens;
- Don't set subpixel order (irrelevant with subpixel rendering disabled)
- Minor refactor

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - ~~x86_64-darwin~~
  - ~~aarch64-darwin~~
- ~~For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))~~
- [x] Tested, manually (no applicable NixOS tests)
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- ~~Tested basic functionality of all binary files (usually in `./result/bin/`)~~
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - ~~(Package updates) Added a release notes entry if the change is major or breaking~~
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - ~~(Module addition) Added a release notes entry if adding a new NixOS module~~
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by nbraud** [comment]
@alyssais, @mkg20001 Could one of you review this, and/or comment on whether a release note is needed?

**Comment by nbraud** [comment]
Hi @RaitoBezarius!

Thanks for taking up the review for this, I'll try to get around to drafting a release note before leaving for Xmas.

**Comment by nbraud** [comment]
Hi @RaitoBezarius !

Sorry it took me so long to get around to this: I didn't manage before leaving for the winter holidays, and was rather ill the last couple months since.

Anyhow, I added an entry to the release notes, so this should hopefully be good to go.

File: nixos/modules/hardware/video/hidpi.nix [hunk 1]
```diff
@@ -12,11 +12,11 @@ with lib;
     boot.loader.systemd-boot.consoleMode = mkDefault "1";
 
 
-    # Grayscale anti-aliasing for fonts
-    fonts.fontconfig.antialias = mkDefault true;
-    fonts.fontconfig.subpixel = {
-      rgba = mkDefault "none";
-      lcdfilter = mkDefault "none";
+    # Disable font anti-aliasing, hinting, and sub-pixel rendering by default
```
Comment (RaitoBezarius) [inline nixos/modules/hardware/video/hidpi.nix:15]: Could I bother you with a link to fontconfig upstream recommendations?

Comment (RaitoBezarius) [reply nixos/modules/hardware/video/hidpi.nix:15]: (as a comment of this file)

File: nixos/modules/hardware/video/hidpi.nix [hunk 2]
```diff
@@ -12,11 +12,11 @@ with lib;
...
+    fonts.fontconfig = {
+      antialias = mkDefault false;
```
Comment (RaitoBezarius) [inline nixos/modules/hardware/video/hidpi.nix:17]: nit: there seems to be a space at the left of this line

**Review by RaitoBezarius — APPROVED** [review]
LGTM modulo link to recos in the file

File: nixos/modules/hardware/video/hidpi.nix (hunk 2 above)
Comment (nbraud) [reply nixos/modules/hardware/video/hidpi.nix:17]: Which line?  I'm not seeing in my usual editor (Emacs with `nix-mode`, enabled `whitespace-mode` to check) nor with `nano` in its default config... but `grep` seems to report some?  ^^'

```shell
$ grep -E '\w$' -- nixos/modules/hardware/video/hidpi.nix
    # Needed when typing in passwords for full disk encryption
    # Disable font anti-aliasing, hinting, and sub-pixel rendering by default
    # TODO Find reasonable defaults X11 & wayland
```

Using `sed` to get rid of it also removed the last character of each of those lines, so I suspect there might be some "funny" codepoints in there, I'll check that shortly  >_>'

Comment (nbraud) [reply nixos/modules/hardware/video/hidpi.nix:17]: That doesn't seem to be Unicode shenanigans: checking the change in that first line with an hex viewer, only the letter `n` (0x6E) is gone, the linebreak is still a single `\n`, then comes the next line's indenting spaces (0x20)

I'm feeling hopelessly confused, so I'll leave it as is and check again tomorrow, but feel free to push a commit to my PR branch fixing the nit.

File: nixos/modules/hardware/video/hidpi.nix (hunk 1 above)
Comment (nbraud) [reply nixos/modules/hardware/video/hidpi.nix:15]: I was referring to the comments on the `fontconfig.subpixel` options [in the NixOS manual](https://nixos.org/manual/nixos/unstable/options.html#opt-fonts.fontconfig.subpixel.lcdfilter):
- `antialias`: "At high resolution (> 200 DPI), antialiasing has no visible effect; users of such displays may want to disable this option."
- `hinting`: "[...] hinting will do nothing (at best);"
- `lcdfilter`: "LCD filtering has no visible effect; users of such displays may want to select `none`."

`fontconfig.nix` does not specify an external reference for those recommendations, and that specific wording does not come up outside NixOS in a web search, but Arch's wiki etc. have similar recommendations.

I can't find something offhand specifically from fontconfig's upstream though, their user manual just covers the configuration syntax.

File: nixos/modules/hardware/video/hidpi.nix (hunk 2 above)
Comment (nbraud) [reply nixos/modules/hardware/video/hidpi.nix:17]: "Mystery" solved: I am being very sleepy and `\w` is not whitespace but word boundaries, and of course POSIX regexp considers those to be characters somehow

... and `rg` confirms there's not trailing whitespace  :sweat_smile:

File: nixos/modules/hardware/video/hidpi.nix (hunk 1 above)
Comment (nbraud) [reply nixos/modules/hardware/video/hidpi.nix:15]: PS: Is it sufficient to link back to the manual?  At least this changes make things internally coherent

**Comment by nbraud** [comment]
@RaitoBezarius Ping?

File: nixos/modules/hardware/video/hidpi.nix (hunk 1 above)
Comment (RaitoBezarius) [reply nixos/modules/hardware/video/hidpi.nix:15]: No needed if it comes from us somewhat, I thought there was a proper fontconfig page somewhere on this, let's mark it resolved

**Comment by RaitoBezarius** [comment]
@nbraud can I ask the release entry commit to have `nixos/hidpi:` as a prefix too?

**Comment by nbraud** [comment]
@RaitoBezarius Done, and added a comment explicitly refering to fontconfig.nix for the choice of font-related settings.

**Comment by qbit** [comment]
FWIW, this broke fonts in KDE for me.

Firefox:
image

Konsole:
image

I had to set `hardware.video.hidpi.enable = false;` to recover the proper rendering.

**Comment by Mic92** [comment]
Here is the revert if you were also affected: https://github.com/Mic92/nixpkgs/commit/9470bbb8034beb6731cf82dc0cf7ab6573f3041d

**Comment by cbergqvist** [comment]
Broke font rendering on Gnome for me as well.

**Comment by RaitoBezarius** [comment]
The fix is being cooked #222236 here.

**Comment by SuperSandro2000** [comment]
Sounds like we want to revert this right now to prevent bigger breakages in nixos-unstable unless the linked PR is merged really soon.

**Comment by jian-lin** [comment]
@RaitoBezarius 
> The fix is being cooked https://github.com/NixOS/nixpkgs/pull/222236 here.

WRT font rendering issues reported here, #222236 is not a fix.  It just renames the option.

I guess changing the default value of `fonts.fontconfig.antialias` causes this issue.  Users affected by this can try to enable antialias.

**Comment by sg-qwt** [comment]
> Sounds like we want to revert this right now to prevent bigger breakages in nixos-unstable unless the linked PR is merged really soon.

Can we pls revert this asap until linked PR is finalized? This is causing huge pain for desktop users on unstable.

**Comment by trofi** [comment]
Commit f03716715f663f1c45056b7df450cf1b7386181b `nixos/hidpi: Disable anti-aliasing` broke font rendering for me. Picture: 


Two windows have different aliasing settings. Guess which is which. The application is an `alacritty` on `sway` on a display with `3840x2160` resolution. The font is `family: Liberation Mono size: 11.5`.

Note how individual pixels stick out of `c` and `m` letters.

I can see very similar font corruption in `firefox` (`Liberation 16` fonts).

Are you sure disabling aliasing is a reasonable thing to do? Should I reconfigure my fonts instead to get rid of it?

**Comment by RaitoBezarius** [comment]
> > Sounds like we want to revert this right now to prevent bigger breakages in nixos-unstable unless the linked PR is merged really soon.
> 
> Can we pls revert this asap until linked PR is finalized? This is causing huge pain for desktop users on unstable.

We merged the other PR now, it should arrive as soon as the channels moves forward.

**Pull request merged**